File: strategies/docker.go
// Copyright 2025 Naren Yellavula
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package strategies

// DockerHelpStrategy handles Docker commands
type DockerHelpStrategy struct {
	cmdRunner *CommandRunner
}

func NewDockerHelpStrategy(cmdRunner *CommandRunner) *DockerHelpStrategy {
	return &DockerHelpStrategy{cmdRunner: cmdRunner}
}

func (d *DockerHelpStrategy) SupportsCommand(baseCmd string) bool {
	return baseCmd == "docker"
}

func (d *DockerHelpStrategy) Priority() int {
	return 2
}

func (d *DockerHelpStrategy) GetHelp(cmdParts []string) (string, error) {
	cmd := NewCommand(cmdParts)

	if !cmd.HasSubCommand(1) {
		return d.cmdRunner.Run("docker", "--help")
	}

	// Handle docker subcommand help
	args := append(cmd.SubCmds, "--help")
	return d.cmdRunner.Run("docker", args...)
}

File: recaller-rs/src/help_system/strategies/docker.rs
use std::sync::Arc;

use anyhow::Result;

use crate::help_system::runner::CommandRunner;
use crate::help_system::strategy::{CommandParts, HelpStrategy};

pub struct DockerHelpStrategy {
    runner: Arc<CommandRunner>,
}

impl DockerHelpStrategy {
    pub fn new(runner: Arc<CommandRunner>) -> Self {
        Self { runner }
    }
}

impl HelpStrategy for DockerHelpStrategy {
    fn priority(&self) -> i32 {
        2
    }

    fn supports_command(&self, base_cmd: &str) -> bool {
        base_cmd == "docker"
    }

    fn get_help(&self, command: &CommandParts) -> Result<String> {
        if !command.has_sub_command(1) {
            return self.runner.run("docker", &["--help"]);
        }

        let mut args: Vec<&str> = command.sub_cmds().iter().map(|s| s.as_str()).collect();
        args.push("--help");
        self.runner.run("docker", &args)
    }
}
